tempesta_fw/ptrset.h

```diff
@@ -0,0 +1,296 @@
+ *
+ * Disadvantages are:
+ *  - Bad asymptotic behavior: add/delete take O(N).
+ *  - Hard limit for a number of pointers in a set. No support for growing.
+ *
+ * Copyright (C) 2012-2014 NatSys Lab. (info@natsys-lab.com).
+ * Copyright (C) 2015 Tempesta Technologies.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#ifndef __TFW_PTRSET_H__
+#define __TFW_PTRSET_H__
+
+#include "lib.h"
+#include "log.h"
+
+/**
+ * Define a custom specific PtrSet structure.
+ *
+ * For example:
+ *     typedef TFW_PTRSET_STRUCT(TfwAddr, 32) TfwMyAddrSet;
+ * will define a type containing an array of 32 items of type "TfwAddr *".
+ *
+ * The fixed maximum number of 32 elements is not always handy.
+ * If you want dynamic allocation, you may specify @static_max_ptrs = 0:
+ *     typedef TFW_PTRSET_STRUCT(TfwAddr, 0) TfwMyAddrSet;
+ * That creates a structure with zero-length array in the tail, so you will be
+ * responsible for memory allocation. Use __tfw_ptrset_max() to calculate size
+ * and __tfw_ptrset_init() to initialize the structure after allocation.
+ */
+#define TFW_PTRSET_STRUCT(ptr_type, static_max_ptrs) \
+struct {					\
+	atomic_t counter;			\
+	short n;				\
+	short max;				\
+	ptr_type *ptrs[static_max_ptrs];	\
+}
+
+/**
+ * Get the maximum number of pointers in a set. Calculate it at compile time for
+ * fixed-length arrays, or use TfwPtrSet->max for variable-length arrays.
+ * That is a little bit faster and allows you to allocate sets statically
+ * without explicit initialization (all the fields are zero'ed which is valid).
+ */
+#define __tfw_ptrset_max(s) \
+	(ARRAY_SIZE((s)->ptrs) ? ARRAY_SIZE((s)->ptrs) : (s)->max)
+
+/**
+ * Calculate total size of a set structure, including the array of pointers.
+ */
+#define tfw_ptrset_size(max) \
+	(sizeof(TfwPtrSet) + ((max) * sizeof(void *)))
+
+/**
+ * Iterate over pointers in a set.
+ *
+ * Usage:
+ *   typedef TFW_PTRSET_STRUCT(TfwAddr, 32) TfwMyAddrSet;
+ *   size_t current_addr_idx;
+ *   TfwAddr *current_addr;
+ *   TfwMyAddrSet addresses;
+ *
+ *   tfw_ptrset_for_each(current_addr, current_addr_idx, addresses) {
+ *           printk("%d\n", current_addr->family);
+ *   }
+ */
+#define tfw_ptrset_for_each(ptr, i, s) \
+	for ((i) = 0; ((i) < (s)->n && (ptr = (s)->ptrs[i])); ++(i))
```

Comment:
Typically foreach is faster for intrusive lists in comparison with such pointer lists.